src/common/events.ts

```diff
@@ -0,0 +1,294 @@
+  eventNamesMap,
+}: {
+  data: Readable | Buffer;
+  dataSize: number;
+  emitter: EventEmitter;
+  eventNamesMap: {
+    'on-progress': string;
+    'on-error': string;
+    'on-end': string;
+  };
+}): Readable {
+  const existingStream = data instanceof Readable ? data : Readable.from(data);
+  const eventingStream = new PassThrough();
+
+  // pause the stream to avoid emitting progress events until the stream is ready
+  existingStream.pause();
+
+  // add listener to emit progress events as the stream is read
+  let processedBytes = 0;
+  existingStream.on('data', (chunk) => {
+    eventingStream.write(chunk);
+    processedBytes += chunk.length;
+    emitter.emit(eventNamesMap['on-progress'], {
+      processedBytes,
+      totalBytes: dataSize,
+    });
+  });
+
+  existingStream.on('end', () => {
+    emitter.emit(eventNamesMap['on-end']);
+    eventingStream.end();
+  });
+
+  existingStream.on('error', (error) => {
+    emitter.emit(eventNamesMap['on-error'], error);
+    eventingStream.destroy(error);
+  });
+
+  // resume the stream to start emitting progress events
+  existingStream.resume();
+  return eventingStream;
+}
+
+/**
+ * Creates an eventing stream from the input data that emits progress and error events
+ */
+export function createStreamWithEvents({
+  data,
+  dataSize,
+  emitter,
+  eventNamesMap,
+}: {
+  data: Readable | Buffer | ReadableStream;
+  dataSize: number;
+  emitter: EventEmitter;
+  eventNamesMap: {
+    'on-progress': string;
+    'on-error': string;
+    'on-end': string;
+  };
+}): Readable | ReadableStream {
+  if (
+    data instanceof ReadableStream ||
+    (typeof window !== 'undefined' && data instanceof Buffer)
+  ) {
+    return createReadableStreamWithEvents({
+      data,
+      dataSize,
+      emitter,
+      eventNamesMap,
+    });
+  }
+
+  if (data instanceof Readable || data instanceof Buffer) {
+    return createReadableWithEvents({
+      data,
+      dataSize,
+      emitter,
+      eventNamesMap,
+    });
+  }
+
+  throw new Error('Invalid data or platform type');
+}
+
+// base class that extends EventEmitter with custom types for events and payloads
+export abstract class TurboEmitter<
+  T extends Record<string, unknown>,
+> extends EventEmitter<Extract<keyof T, string>> {
+  override on<K extends keyof T>(
+    event: K,
+    listener: (...args: any[]) => void,
+  ): this {
+    // @ts-expect-error - TODO: eventemitter3 has strict types
+    return super.on(event, listener);
+  }
+
+  override emit<K extends keyof T>(event: K, ...args: any[]): boolean {
+    // @ts-expect-error - TODO: eventemitter3 has strict types
+    return super.emit(event, ...args);
+  }
+}
+
+export class UploadEmitter extends TurboEmitter<TurboUploadEventsAndPayloads> {
+  constructor({
+    onUploadProgress,
+    onUploadError,
+    onUploadSuccess,
+  }: TurboUploadEmitterEventArgs = {}) {
+    super();
+    if (onUploadProgress !== undefined) {
+      this.on('upload-progress', onUploadProgress);
```

Comment:
should these be `on-*` now?

Reply:
i prefer it verbose for now - [chore(test): increase test coverage and add assertions for signing an…](https://github.com/ardriveapp/turbo-sdk/pull/254/commits/4761bca1c2aeacefdd1c9aee65f2bc858b488092)